camel/utils/python_interpreter.py

```diff
@@ -0,0 +1,359 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import ast
+import difflib
+import importlib
+from typing import Any, Dict, List, Mapping, Optional, Union
+
+
+class InterpreterError(ValueError):
+    r"""An error raised when the interpreter cannot evaluate a Python
+    expression, due to syntax error or unsupported operations.
+    """
+
+    pass
+
+
+class PythonInterpreter():
+    r"""A customized python interpreter to control the execution of
+    LLM-generated codes. The interpreter makes sure the code can only execute
+    functions given in action space and import white list. It also supports
+    fuzzy variable matching to reveive uncertain input variable name.
+
+    Args:
+        action_space (Dict[str, Any]): The dictionary mapping action names to
+            the correponding functions or classes.
+        import_white_list (List[str], optional): The list storing the
+            importable python modules or functions which can be imported in
+            code. All submodules and functions of the modules in the list are
+            importable. Any other import statements will be rejected. :obj:`.`
+            is the seperator between the module and its submodule or its
+            function name. (default: :obj:`[]`)
+    """
+
+    def __init__(self, action_space: Dict[str, Any],
+                 import_white_list: List[str] = []) -> None:
```

Comment:
Default empty list initialization is an error: https://stackoverflow.com/questions/366422/how-can-i-avoid-issues-caused-by-pythons-early-bound-default-parameters-e-g-m